app/gui/src/project-view/components/GraphEditor/widgets/WidgetTableEditor/editHandler.ts

```diff
@@ -0,0 +1,166 @@
+import { NEW_COLUMN_ID } from '@/components/GraphEditor/widgets/WidgetTableEditor/tableInputArgument'
+import { WidgetEditHandler, WidgetEditHooks } from '@/providers/widgetRegistry/editHandler'
+import { ToValue } from '@/util/reactivity'
+import { CellPosition, StartEditingCellParams } from 'ag-grid-enterprise'
+import { computed, ref, toValue, watch } from 'vue'
+
+export interface EditedCell {
+  rowIndex: number | 'header'
+  colKey: string
+}
+
+/**
+ * A composable containing logic for WidgetTableEditor's edit handler.
+ *
+ * Contains logic of synchronizing state between AgGrid cells and our custom headers,
+ * and react for user input received from them.
+ */
+export function useTableEditHandler(
+  gridApi: ToValue<
+    | {
+        stopEditing(cancel: boolean): void
+        startEditingCell(editedCell: StartEditingCellParams): void
+        getEditingCells(): Array<CellPosition>
+      }
+    | undefined
+  >,
+  colDefs: ToValue<{ colId: string }[]>,
+  widgetHandlerConstructor: (hooks: WidgetEditHooks) => WidgetEditHandler,
+) {
+  const columnIndexById = computed(
+    () => new Map(toValue(colDefs).map((col, index) => [col.colId, index])),
+  )
+  const firstColumn = computed(() => toValue(colDefs)[1]) // The 0 col is un-editable row index.
+
+  const editedCell = ref<EditedCell>()
+  let revertChangesCb: (() => void) | undefined
+  function syncGridWithEditedCell(cell = editedCell.value) {
+    const api = toValue(gridApi)
+    if (!api) return
+    const editedInGrid = api.getEditingCells()[0]
+    if (cell == null || cell.rowIndex === 'header') {
+      api.stopEditing(false)
+    } else if (
+      editedInGrid?.rowIndex !== cell.rowIndex ||
+      editedInGrid?.column.getColId() !== cell.colKey
+    ) {
+      api.startEditingCell({ rowIndex: cell.rowIndex, colKey: cell.colKey })
+    }
+  }
+  watch(editedCell, (cell) => {
+    syncGridWithEditedCell(cell)
+    if (cell != null && !handler.isActive()) {
+      handler.start()
+    } else if (cell == null && handler.isActive()) {
+      handler.end()
+    }
+  })
+
+  const handler = widgetHandlerConstructor({
+    cancel() {
+      revertChangesCb?.()
+      editedCell.value = undefined
+    },
+    end() {
+      editedCell.value = undefined
+    },
+    suspend: () => {
+      return {
+        resume: () => syncGridWithEditedCell(),
+      }
+    },
```

Comment:
Does this do anything? A `suspend` implementation that doesn't capture any state is suspicious.